### PR TITLE
feat(web-components): add tablist

### PR DESCRIPTION
**Pull request by davatron5000**
## Previous Behavior

Current Tabs web component is an assembly of three components (`-tabs`, `-tab`, and `-tabpanel`) which is very explicit but very DOM heavy and doesn't easily support scenarios like ajax'ing in content.

## New Behavior

This new `Tablist` component is more on par with the React implementation and is much simpler in its approach, providing only the`-tab` and `-tablist` components and allowing the consumer to react to the `change` event (when the active tab changes) as necessary.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

✅ No changes found
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentui-web-components-v3 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_web-components"/>

File: packages/web-components/src/tablist/tablist.styles.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+export const styles = css`
+  ${display('flex')}
+
+  :host {
+    --tabPaddingInline: inherit;
+    --tabPaddingBlock: inherit;
+    --tabIndicatorInsetInline: 0;
+    --tabIndicatorInsetBlock: 0;
+
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.styles.ts:36]: NIT: extra space

File: packages/web-components/src/tablist/tablist.styles.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+  :host ::slotted([role='tab'][data-animate='true'])::after {
+    transition-property: transform;
+    transition-duration: ${durationSlow};
+    transition-timing-function: ${curveDecelerateMax};
+  }
+
+  :host ::slotted([role='tab'])::after {
+    height: ${strokeWidthThicker};
+    margin-block-start: auto;
+    transform-origin: left;
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.styles.ts:60]: Will this conflict with document direction?

File: packages/web-components/src/tablist/tablist.stories.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    <fluent-tab> Third Tab</fluent-tab>
+    <fluent-tab> Fourth Tab</fluent-tab>
+  </fluent-tabs>
+
+  <fluent-tablist>
+    <fluent-tab> First Tab </fluent-tab>
+    <fluent-tab disabled> Second Tab</fluent-tab>
+    <fluent-tab> Third Tab</fluent-tab>
+    <fluent-tab> Fourth Tab</fluent-tab>
+  </fluent-tabs>
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.stories.ts:77]: wrong closing tag

File: packages/web-components/src/tablist/tablist.stories.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+`;
+export const TablistAppearance = renderComponent(tabsAppearance).bind({});
+
+const tabsDisabledTablist = html`
+  <fluent-tablist disabled>
+    <fluent-tab> First Tab </fluent-tab>
+    <fluent-tab> Second Tab</fluent-tab>
+    <fluent-tab> Third Tab</fluent-tab>
+    <fluent-tab> Fourth Tab</fluent-tab>
+  </fluent-tabs>
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.stories.ts:70]: wrong closing tag

File: packages/web-components/src/tablist/tablist.stories.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+`;
+export const TablistDefault = renderComponent(tabsDefault).bind({});
+
+const tabsHorizontal = html`
+  <fluent-tablist orientation="horizontal">
+    <fluent-tab> First Tab </fluent-tab>
+    <fluent-tab> Second Tab</fluent-tab>
+    <fluent-tab> Third Tab</fluent-tab>
+    <fluent-tab> Fourth Tab</fluent-tab>
+  </fluent-tabs>
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.stories.ts:34]: Same comment a bunch of these have the wring closing tag

File: packages/web-components/src/tablist/tablist.template.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { html, slotted } from '@microsoft/fast-element';
+import { Tablist } from './tablist.js';
+
+export const template = html<Tablist>`
+  <template role="tablist">
+    <slot name="tab" ${slotted('tabs')}></slot>
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.template.ts:6]: I think the old tabs had a start and end slot. Might no longer be required but pretty sure it had those slots.

File: packages/web-components/src/tablist/tablist.ts [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+   */
+  public connectedCallback(): void {
+    super.connectedCallback();
+
+    this.tabIds = this.getTabIds();
+    this.activeTabIndex = this.getActiveIndex();
+  }
+}
+
+export class Tablist extends BaseTablist {
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.ts:283]: Nice, thanks for doing this.

File: packages/web-components/src/tablist/tablist.ts [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+   */
+  @attr
+  public appearance?: TablistAppearance = TablistAppearance.transparent;
+
+  /**
+   * disabled
+   * Used for disabling all click and keyboard events for the tabs, child tab elements and tab panel elements. UI styling of content and tabs will appear as "grayed out."
+   */
+  @attr({ mode: 'boolean' })
+  public disabled?: boolean;
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.ts:317]: Should disabled be in the base class?

File: packages/web-components/src/tablist/tablist.ts [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+        this.previousActiveTabData?.x !== this.activeTabData?.x &&
+        this.previousActiveTabData?.y !== this.activeTabData?.y
+      ) {
+        this.previousActiveTabData = this.activeTabData;
+      }
+    }
+  }
+
+  private setTabOffsetCSSVar() {
+    this.style.setProperty('--tabIndicatorOffset', `${this.activeTabOffset}px`);
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.ts:418]: SO, its preferrable to use the `$FASTController.addStyles()` instead of `style.setPorperty`.  Add Styles API will add the styles to the host element instead of to the styles property on the element itself. Also if someone set inline `styles` on this element this would collide.

```
private myStyles: ElementStyles | undefined;

private setTabOffsetCSSVar() {
   this.myStyles = css`
        :host {
          --tabIndicatorOffset: `${this.activeTabOffset}px`;
        }
      `;
      this.$fastController.addStyles(this.myStyles);
}
```

You can call removeStyles in the disconnectedCallback.

File: packages/web-components/src/tablist/tablist.template.ts (hunk 6 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.template.ts:6]: It did have start/end slots. Happy to add them back, but my thinking was that because this is now more composable you wouldn't need them in the tab control itself.

```html
<div class="flex">
  <button>my before content</button>
  <fluent-tablist>...</fluent-tablist>
  <button>my after content</button>
</div>
```

File: packages/web-components/src/tablist/tablist.ts [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+    }
+  };
+
+  /**
+   * The adjust method for FASTTabs
+   * @public
+   * @remarks
+   * This method allows the active index to be adjusted by numerical increments
+   */
+  public adjust(adjustment: number): void {
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.ts:203]: Do we need `adjust`, `adjustForward` and `adjustBackward`. can the latter be achieved by calling `adjust(1)` and `adjust(-1)`?

**Review by eljefe223 — APPROVED** [review]

File: packages/web-components/src/tablist/tablist.template.ts (hunk 6 above)
Comment (eljefe223) [reply packages/web-components/src/tablist/tablist.template.ts:6]: Cool I actually like that way better. carry on.

File: packages/web-components/src/tablist/tablist.styles.ts (hunk 2 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.styles.ts:60]: Logical properties don't work with `transform-origin`. I'll need to spin up another story to verify RTL.

File: packages/web-components/src/tablist/tablist.ts [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+    switch (event.key) {
+      case keyHome:
+        event.preventDefault();
+        this.adjust(-this.activeTabIndex);
+        break;
+      case keyEnd:
+        event.preventDefault();
+        this.adjust(this.tabs.length - this.activeTabIndex - 1);
+        break;
+    }
```
Comment (chrisdholt) [inline packages/web-components/src/tablist/tablist.ts:194]: I wonder if it would be slightly more efficient to combine these into one switch statement and check within the case? In that scenario we're accessing event.key once and catching in the switch once. Example - If we're horizontal we can eject from up/down:

```ts
case keyArrowUp:
    if (this.isHorizontal()) { return }
    event.preventDefault();
    this.adjustBackward(event);
    break;
```

**Comment by chrisdholt** [comment]
General question - should our stories provide examples on how to wire up panels?

**Comment by davatron5000** [comment]
> Should we be removing old / unused tabs and tab panel as part of this PR?

@eljefe223 I talked with @chrisdholt about this yesterday and old-tabs will probably be removed/deprecated but it's probably worth doing outside of this commit since this is additive (non-breaking change) and removing tabs would potentially be a breaking change for some consumers.

> General question - should our stories provide examples on how to wire up panels?

I think we probably should. Storybook in its current form doesn't really allow us to show event handlers and extra script, but I'll see what I can do.

File: packages/web-components/src/tablist/tablist.styles.ts (hunk 2 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.styles.ts:62]: Verified that everything works as expected in RTL. Will have a story up for it soon. I think we actually lucked out a bit because `--tabIndicatorOffset` uses `getBoundingClientRect.x` which would also be directional-based.

**Comment by davatron5000** [comment]
RTL story and fully wired tabpanel story added to primary story in 10dc1f7fb1.

To get these all working I needed to update the tablist class and base class...
- `$emit('change')` when its done setting up the tabs  
- Run the animation loop after the tabsChanged change handler to set the correct positioning for the active animation. (Without doing this the animation would start at the far left)
- Update the left/right arrows to be `direction` sensitive. This matches the React implementation.

And fwiw, the React implementation infinitely loops the roving index whereas the web component implementation stops at the end of the list. IIRC, looping back to the beginning/end is the expected behavior. This change could happen now or be an improvement later on if needed. Let me know what to do.

File: packages/web-components/src/tablist/tablist.ts (hunk 9 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.ts:388]: One issue I'm seeing with generating a sheet is that it's creating a stylesheet every time a tab is clicked (see attached). To me this seems... I dunno... wasteful? Maybe memory intensive? 

Screenshot 2024-07-24 at 9 31 32 PM

I'm open to the change but I guess I'd like to know more about the concern that someone is going to wholesale wipeout `element.style`. When I use `setProperty` it self-repairs on the next click.

Comment (chrisdholt) [reply packages/web-components/src/tablist/tablist.ts:388]: A couple things, and this is a good convo to have IMO...

I think we want to call `.removeStyles()` before we set new styles in the method as that will avoid this continually adding of new sheets. Essentially, wipe the sheet, create the sheet.

With that said, I think that using `.addStyles()` is often preferrable but not always ideal. Sometimes it's super clear - as I'm exploring tooltip, until we get the `anchor` attribute we need to set it via CSS. It's _likely_ that the reference doesn't change often with how I'm constructing it and so I'll _probably_ use the `.addStyles()` API when `anchor` isn't available. With that said, in something like Slider, the calls to destroying and creating a new stylesheet are significant when dragging the slider and trying to update a CSS Variable value, it's happening constantly. In this case, I think the creation and removal of stylesheets is likely a recipe for trouble.

So, in this scenario, we should consider the same things - what's best for the situation, what are the possible issues with it (overriding inline styles, performance, etc) and come to what we think is best. Ultimately, this is "internals" so we can learn and iterate from it.

**Comment by chrisdholt** [comment]
> RTL story and fully wired tabpanel story added to primary story in [10dc1f7](https://github.com/microsoft/fluentui/commit/10dc1f7fb158cf1754eaf2ac2e516da8f22b5177).
> 
> To get these all working I needed to update the tablist class and base class...
> 
> * `$emit('change')` when its done setting up the tabs
> * Run the animation loop after the tabsChanged change handler to set the correct positioning for the active animation. (Without doing this the animation would start at the far left)
> * Update the left/right arrows to be `direction` sensitive. This matches the React implementation.
> 
> And fwiw, the React implementation infinitely loops the roving index whereas the web component implementation stops at the end of the list. IIRC, looping back to the beginning/end is the expected behavior. This change could happen now or be an improvement later on if needed. Let me know what to do.

@davatron5000 Let's look to wire up the keyboard behavior to properly loop. There's likely a couple ways this could be done. There is a utility from `@microsoft/fast-web-utilities`, `wrapInBounds()` which might make it relatively simple: https://www.npmjs.com/package/@microsoft/fast-web-utilities#wrapinbounds

File: packages/web-components/src/tablist/tablist.styles.ts [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+    --tabPaddingInline: inherit;
+    --tabPaddingBlock: inherit;
+    --tabIndicatorInsetInline: 0;
+    --tabIndicatorInsetBlock: 0;
+    box-sizing: border-box;
+    color: ${colorNeutralForeground2};
+    flex-direction: row;
+  }
+
+  :host([orientation='vertical']) {
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.styles.ts:44]: Should we use CSS partials from `src/styles/states/index.ts` for orientation and size?

File: packages/web-components/src/tablist/tablist.styles.ts [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+    flex-direction: column;
+  }
+
+  :host ::slotted([role='tab']) {
+    align-items: flex-start;
+  }
+
+  /* indicator animation  */
+  :host ::slotted([role='tab'][data-animate='true'])::after {
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.styles.ts:53]: I wonder if we should be specific about the tab element type, i.e. instead of using `[role='tab']`, use `fluent-tab`, because if an author decides to use a different element, this may break their element's styles.

File: packages/web-components/src/tablist/tablist.ts [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+ */
+export class BaseTablist extends FASTElement {
+  /**
+   * Used for disabling all click and keyboard events for the tabs, child tab elements.
+   * @public
+   * @remarks
+   * HTML Attribute: disabled.
+   */
+  @attr({ mode: 'boolean' })
+  public disabled?: boolean;
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.ts:29]: Should we assign a default value as `false`?

File: packages/web-components/src/tablist/tablist.ts [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+   * @public
+   * @remarks
+   * HTML Attribute: orientation
+   */
+  @attr
+  public orientation: TablistOrientation = TablistOrientation.horizontal;
+  /**
+   * @internal
+   */
+  public orientationChanged(): void {
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.ts:42]: Should all the changed methods be `protected` instead of `public`?

File: packages/web-components/src/tablist/tablist.ts [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+    if (this.activeTabIndex !== this.prevActiveTabIndex) {
+      this.activeid = this.tabIds[this.activeTabIndex] as string;
+      this.focusTab();
+      this.change();
+    }
+  }
+
+  private handleTabClick = (event: MouseEvent): void => {
+    const selectedTab = event.currentTarget as HTMLElement;
+    if (selectedTab.nodeType === 1 && this.isFocusableElement(selectedTab)) {
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.ts:159]: Nit: For better code readability, `selectedTab.nodeType === Node.ELEMENT_NODE`.

File: packages/web-components/src/tablist/tablist.ts [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+
+  private prevActiveTabIndex: number = 0;
+  private activeTabIndex: number = 0;
+  private tabIds!: Array<string>;
+
+  private change = (): void => {
+    this.$emit('change', this.activetab);
+  };
+
+  private isDisabledElement = (el: Element): el is HTMLElement => {
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.ts:95]: Can we move these utility functions out of the class so that they won't be added to every instance?

File: packages/web-components/src/tablist/tablist.ts [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+    }
+  }
+
+  /**
+   * A reference to the active tab
+   * @public
+   */
+  public activetab!: HTMLElement;
+
+  private prevActiveTabIndex: number = 0;
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.ts:87]: Have you tried to store the elements instead of index? The index method can be fragile if an element gets removed or moved.

File: packages/web-components/src/tablist/tablist.ts [hunk 19]
```diff
@@ -0,0 +1,408 @@
...
+      this.setComponent();
+    }
+  };
+
+  private isHorizontal(): boolean {
+    return this.orientation === TablistOrientation.horizontal;
+  }
+
+  private handleTabKeyDown = (event: KeyboardEvent): void => {
+    const dir = window.getComputedStyle(this).direction;
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.ts:171]: You can use the utility function `getDirection()` in `src/utils/direction.js`.

File: packages/web-components/src/tablist/tablist.ts [hunk 20]
```diff
@@ -0,0 +1,408 @@
...
+
+    // the index of the next focusable tab within the context of all available tabs
+    const nextIndex = this.tabs.indexOf(focusableTabs[nextTabIndex]);
+
+    if (nextIndex > -1) {
+      this.moveToTabByIndex(this.tabs, nextIndex);
+    }
+  }
+
+  private moveToTabByIndex(group: HTMLElement[], index: number) {
```
Comment (marchbox) [inline packages/web-components/src/tablist/tablist.ts:232]: Nit: rename this to `activateTabByIndex`?

File: packages/web-components/src/tablist/tablist.styles.ts (hunk 13 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: In the `fluent-tabs` it is `fluent-tab` but my reasoning for something more generic was that consumers might recompose `fluent-tab` with a different prefix (I saw `fluent-beta-` the other day). So using `role` (hopefully) makes it a bit more agnostic to the component registration. Open to ideas here.

Comment (chrisdholt) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: I think the only concern I have here is that when this goes to element internals the role will not reflect. Agree on the flexibility here, I wonder...does it work to reference `FluentDesignSystem.prefix`? I don't think we've gotten that addressed to allow teams to change the prefix properly yet, but that _could_ in theory work?

Comment (marchbox) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: Ah I didn't know there could be `fluent-beta-` elements. In that case, maybe `::slotted([slot="tab"])` can solve Chris' concern over `role` not being exposed due to element internals.

Comment (chrisdholt) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: I think the tension here is that we would need to then have these _not_ be the default slot, which would be weird ergonomically as the tabs are the most intuitive and obvious content to be expected as the default slots.

Perhaps, given that the outliers would be `start` and `end` slots, we look at `*:not()`? Not ideal, but scoped within the shadow root and not _every_ element?

Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: Some good news maybe, I removed the `start/end` slots on tablist, so we could maybe just do `::slotted(*)`. Unless we need `start/end`.. then it's bad news.

Comment (marchbox) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: I was trying to understand this earlier but haven't figured it out: if the tab elements are assigned to the default slot, how does this work? https://github.com/microsoft/fluentui/blob/1931b54c1517fae68a004022fad736c1c06168de/packages/web-components/src/tablist/tablist.ts#L126-L142

Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: So `slot="tab"` and `role="tab"` get set by the `Tab` component and `Tablist` does not have a default slot, only `<slot name="tab">`.

https://github.com/microsoft/fluentui/blob/1931b54c1517fae68a004022fad736c1c06168de/packages/web-components/src/tab/tab.template.ts#L7

Comment (marchbox) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: Ah I see. I was confused why there was no `slot="tab"` on the `<fluent-tab>` elements in the stories 😄

Comment (chrisdholt) [reply packages/web-components/src/tablist/tablist.styles.ts:53]: Ah, so `::slotted([slot="tab"]) {}`

File: packages/web-components/src/tablist/readme.md [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+| --------------- | ------------------- |
+| vertical        | orientation         |
+| selected-value  | activeid            |
+| value           | id                  |
+
+## Implementation - Sample Code
+
+### Default
+
+By default Tablists are arranged horizontally. The developer sets `activeid` Tablist attribute. The Component handles the logic of what is shown and hidden when user clicks on the tabs. For switcing to work correctly the tab list requires that the indexing of the tabs and tab-panels be organized to correspond to their matching items - the order of the tabs must match the order of the tab panels:
```
Comment (eljefe223) [inline packages/web-components/src/tablist/readme.md:70]: spelling "switcing"

File: packages/web-components/src/tablist/tablist.ts [hunk 22]
```diff
@@ -0,0 +1,454 @@
...
+   * Used to scale the tab active indicator up or down as animations of the active indicator occur.
+   */
+  private activeTabScale = 1;
+
+  /**
+   * appearance
+   * There are two modes of appearance: transparent and subtle.
+   */
+  @attr
+  public appearance?: TablistAppearance = TablistAppearance.transparent;
```
Comment (eljefe223) [inline packages/web-components/src/tablist/tablist.ts:309]: Do we want to abstract a base class?

File: packages/web-components/src/tablist/tablist.ts (hunk 9 above)
Comment (eljefe223) [reply packages/web-components/src/tablist/tablist.ts:433]: I added the check above in my code sample that checks whether the styles are defined and if they are removes them before it tries to set them again. I just forgot to add that bit.

File: packages/web-components/src/tablist/tablist.ts (hunk 22 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.ts:301]: So that is already abstracted out of the base class. 

```
export class BaseTablist extends FASTElement {
    activeid: string;
    protected activeidChanged(oldValue: string, newValue: string): void;
    activetab: HTMLElement;
    adjust(adjustment: number): void;
    connectedCallback(): void;
    disabled: boolean;
    protected disabledChanged(prev: boolean, next: boolean): void;
    elementInternals: ElementInternals;
    orientation: TablistOrientation;
    protected orientationChanged(prev: TablistOrientation, next: TablistOrientation): void;
    protected setTabs(): void;
    tabs: HTMLElement[];
    protected tabsChanged(): void;
}
```

And then class extension looks like this... (handles the appearance, size, and custom indicator behaviors)

```
export class Tablist extends BaseTablist {
    activeidChanged(oldValue: string, newValue: string): void;
    appearance?: TablistAppearance;
    protected appearanceChanged(prev: TablistAppearance, next: TablistAppearance): void;
    size?: TablistSize;
    protected sizeChanged(prev: TablistSize, next: TablistSize): void;
    tabsChanged(): void;
}
```

File: packages/web-components/src/tablist/tablist.ts (hunk 17 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.ts]: Created a `src/utils/focusable-elements.ts` file. Let me know if that looks good.

File: packages/web-components/src/tablist/tablist.ts (hunk 22 above)
Comment (eljefe223) [reply packages/web-components/src/tablist/tablist.ts:301]: Also super weird... because I swore I only saw Tablist class.

**Review by eljefe223 — APPROVED** [review]
pending any open comments being resolved.

File: packages/web-components/src/tablist/tablist.ts (hunk 9 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.ts:433]: After some perf investigation, the delta here is pretty slim on the performance angle but `setProperty` consistently pulls ahead slightly. In the grandparent method the variables get set before & after setting the active tab so we get two calls per animation loop....

| | Before Call | After Call |
|--------|--------|--------|
| addStyles | 0.1-0.3ms | 0.1-0.3ms |
| setProperty  | 0.1-0.3ms | 0ms |

At 20x CPU slowdown that upper bounds goes to ~4ms (dropped frame territory) but `setProperty` still manages to hit 0ms more often. 

The bigger story might be how much work the browser is doing...

addStyles:
Screenshot 2024-07-31 at 1 28 14 PM

setProperty:
Screenshot 2024-07-31 at 1 29 40 PM

 You can see the flame chart for `addStyles` is a bit deeper of a callstack than for `setProperty`. Which is somewhat expected, but less work is probably better. `addStyles` also increased the surface area API of our tablist class with more imports and a `disconnectedCallback` to cleanup, which I believe one of our stated goals is to have a narrow API surface. 
 
tl;dr– No big alarm bells are going off for me with either approach, but perhaps in the interest of simplicity and marginal gains we go with `setProperty` for now. `addStyles` is essentially a cosmetic DX improvement where you don't see the `style="--foo; --bar;"` attribute on the element in DevTools but get a constructed stylesheet in the styles panel instead. 

If we decide that DX is worth it we can always reintroduce `addStyles` and we have a little bit more insight into the cost/benefits. ❤️

Comment (chrisdholt) [reply packages/web-components/src/tablist/tablist.ts:433]: Thanks for digging in here - there is a lot of really good data. I'm in favor of rolling with setProperty to kick off.

File: packages/web-components/src/tablist/tablist.ts (hunk 18 above)
Comment (davatron5000) [reply packages/web-components/src/tablist/tablist.ts:117]: Discussed this issue with @marchbox. tl;dr– because `setTabs` will re-instantiate the tablist on changes to `this.tabs`, the `prevActiveTabIndex` is a fairly shallow mechanism for storing the previous tab's index in the array of tabs when the tabs are click/keyboard interacted. The state is backed up and "saved" in the `<*-tablist activeid>` attribute which also sets `prevActiveTabIndex` on `activeidChanged`, so while brittle it should rescue successfully. 

Resolving this conversation for now.

**Review by marchbox — APPROVED** [review]
